python/dgl/distributed/role.py

```diff
@@ -0,0 +1,200 @@
+        # Clients are blocked util all clients register their roles.
+        if total_count == kv_store.num_clients:
+            res_list = []
+            for target_id in range(kv_store.num_clients):
+                print('send', target_id)
+                res_list.append((target_id, RegisterRoleResponse(REG_ROLE_MSG)))
+            return res_list
+        return None
+
+GET_ROLE = 700002
+GET_ROLE_MSG = "Get_Role"
+
+class GetRoleResponse(rpc.Response):
+    """Send the roles of all client processes"""
+    def __init__(self, role):
+        self.role = role
+        self.msg = GET_ROLE_MSG
+
+    def __getstate__(self):
+        return self.role, self.msg
+
+    def __setstate__(self, state):
+        self.role, self.msg = state
+
+class GetRoleRequest(rpc.Request):
+    """Send a request to get the roles of all client processes."""
+    def __init__(self):
+        self.msg = GET_ROLE_MSG
+
+    def __getstate__(self):
+        return self.msg
+
+    def __setstate__(self, state):
+        self.msg = state
+
+    def process_request(self, server_state):
+        kv_store = server_state.kv_store
+        return GetRoleResponse(kv_store.role)
+
+# The key is role, the value is a dict of mapping RPC rank to a rank within the role.
+PER_ROLE_RANK = {}
+
+# The global rank of a client process. The client processes of the same role have
+# global ranks that fall in a contiguous range.
+GLOBAL_RANK = {}
+
+# The role of the current process
+CUR_ROLE = None
+
+def init_role(role):
+    """Initialize the role of the current process.
+
+    Each process is associated with a role so that we can determine what
+    function can be invoked in a process. For example, we do not allow some
+    functions in sampler processes.
+
+    The initialization includes registeration the role of the current process and
+    get the roles of all client processes. It also computes the rank of all client
+    processes in a deterministic way so that all clients will have the same rank for
+    the same client process.
+    """
+    global CUR_ROLE
+    CUR_ROLE = role
+    # Register the current role. This blocks until all clients register themselves.
+    client_id = rpc.get_rank()
+    machine_id = rpc.get_machine_id()
+    request = RegisterRoleRequest(client_id, machine_id, role)
+    rpc.send_request(0, request)
+    response = rpc.recv_response()
+    assert response.msg == REG_ROLE_MSG
+
+    # Get all clients on all machines.
+    request = GetRoleRequest()
+    rpc.send_request(0, request)
+    response = rpc.recv_response()
+    assert response.msg == GET_ROLE_MSG
+
+    # Here we want to compute a new rank for each client.
+    # We compute the per-role rank as well as global rank.
+    # For per-role rank, we ensure that all ranks within a machine is contiguous.
+    # For global rank, we also ensure that all ranks within a machine are contiguous,
+    # and all ranks within a role are contiguous.
+    global PER_ROLE_RANK
+    global GLOBAL_RANK
```

Comment:
Consiter put these two into RPCContext.

Reply:
these are fairly complex data structures. i'm not sure if role affects the behavior of RPC.